compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1428,9 +1429,22 @@ private BType checkListConstructorCompatibility(BType bType, BLangListConstructo
                 if (memCompatibiltyType != symTable.semanticError && dlog.errorCount() == 0 &&
                         isUniqueType(compatibleTypes, memCompatibiltyType)) {
                     compatibleTypes.add(memCompatibiltyType);
+                } else if (memCompatibiltyType == symTable.semanticError &&
+                        listCompatibleMemType.tag == TypeTags.ARRAY &&
+                        ((BArrayType) listCompatibleMemType).state == BArrayState.INFERRED) {
+                    this.dlog.unmute();
+                    dlog.error(listCompatibleMemType.tsymbol.pos, DiagnosticErrorCode.CANNOT_INFER_ARRAY_SIZE,
+                            listCompatibleMemType);
+                    this.dlog.mute();
+                    cannotInferArraySize = true;
                 }
             }
 
+            if (cannotInferArraySize) {
+                this.dlog.unmute();
```

Comment:
Since this method is called recursively, we need to fall back to the previous error logging state and the error count. If we were in mute previously, then we should not unmute here. We can let L1448-L1452 handle it. 

I think we can get rid of `cannotInferArraySize` flag.  As we log an error intentionally at L1436 we can do a `errorCount++` there to get the correct count.